src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -0,0 +1,1449 @@
+                        break;
+                    }
+
+                    _api.textProps[a.textLength - 1].canBreakShapingAfter = 1;
+
+                    size_t beg = 0;
+                    for (size_t i = 0; i < a.textLength; ++i)
+                    {
+                        if (_api.textProps[i].canBreakShapingAfter)
+                        {
+                            _emplaceGlyph(mappedFontFace.get(), scale, a.textPosition + beg, a.textPosition + i + 1);
+                            beg = i + 1;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void AtlasEngine::_emplaceGlyph(IDWriteFontFace* fontFace, float scale, size_t bufferPos1, size_t bufferPos2)
+{
+    // This would seriously blow us up otherwise.
+    Expects(bufferPos1 < bufferPos2 && bufferPos2 <= _api.bufferLine.size());
+
+    const auto chars = &_api.bufferLine[bufferPos1];
+    const auto charCount = bufferPos2 - bufferPos1;
+    // _flushBufferLine() ensures that bufferLineColumn.size() > bufferLine.size().
+    const auto x1 = _api.bufferLineColumn[bufferPos1];
+    const auto x2 = _api.bufferLineColumn[bufferPos2];
+
+    Expects(x1 < x2 && x2 <= _api.cellCount.x);
+
+    const u16 cellCount = x2 - x1;
+
+    auto attributes = _api.attributes;
+    attributes.cellCount = cellCount;
+
+    const auto [it, inserted] = _r.glyphs.emplace(std::piecewise_construct, std::forward_as_tuple(attributes, gsl::narrow<u16>(charCount), chars), std::forward_as_tuple());
+    const auto& key = it->first;
+    auto& value = it->second;
+
+    if (inserted)
+    {
+        // Do fonts exist *in practice* which contain both colored and uncolored glyphs? I'm pretty sure...
+        // However doing it properly means using either of:
+        // * IDWriteFactory2::TranslateColorGlyphRun
+        // * IDWriteFactory4::TranslateColorGlyphRun
+        // * IDWriteFontFace4::GetGlyphImageData
+        //
+        // For the first two I wonder how one is supposed to restitch the 27 required parameters from a
+        // partial (!) text range returned by GetGlyphs(). Our caller breaks the GetGlyphs() result up
+        // into runs of characters up until the first canBreakShapingAfter == true after all.
+        // There's no documentation for it and I'm sure I'd mess it up.
+        // For very obvious reasons I didn't want to write this code.
+        //
+        // IDWriteFontFace4::GetGlyphImageData seems like the best approach and DirectWrite uses the
+        // same code that GetGlyphImageData uses to implement TranslateColorGlyphRun (well partially).
+        // However, it's a heavy operation and parses the font file on disk on every call (TranslateColorGlyphRun doesn't).
+        // For less obvious reasons I didn't want to write this code either.
+        //
+        // So this is a job for future me/someone.
+        // Bonus points for doing it without impacting performance.

```

Comment:
LOL. Good luck future person. I wonder if we couldn't get the DWrite folks to offer a better way at this stuff through a post 1.0 version of Project Reunion/DWriteCore.

Reply:
They trivially could. This information is readily available internally.
But it wouldn't make sense with the current API, which is entirely centered around glyph runs.
Maybe you could assembly a fake glyph run and only set some of the fields and hope it doesn't break dwrite internally?